topdown/trace_test.go

```diff
@@ -961,3 +961,72 @@ func TestBufferTracerTraceConfig(t *testing.T) {
 		t.Fatalf("Expected config: %+v, got %+v", expected, conf)
 	}
 }
+
+func TestTracePlug(t *testing.T) {
+	ctx := context.Background()
+	module := `
+		package test
+
+		rule[[a, b]] {
+			a = [1, 2][_]
+			b = [2, 1][_]
+		}
+	`
+
+	compiler := compileModules([]string{module})
+	queryCompiler := compiler.QueryCompiler()
+
+	compiledQuery, err := queryCompiler.Compile(ast.MustParseBody("data.test.rule"))
+	if err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+
+	store := inmem.New()
+	txn := storage.NewTransactionOrDie(ctx, store)
+	defer store.Abort(ctx, txn)
+
+	tracer := plugRuleHeadKeyRecorder{}
+	query := NewQuery(compiledQuery).
+		WithQueryCompiler(queryCompiler).
+		WithCompiler(compiler).
+		WithStore(store).
+		WithTransaction(txn).
+		WithQueryTracer(&tracer)
+
+	if _, err := query.Run(ctx); err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+
+	expected := []ast.Value{
+		ast.NewArray(ast.NumberTerm("1"), ast.NumberTerm("2")),
+		ast.NewArray(ast.NumberTerm("1"), ast.NumberTerm("1")),
+		ast.NewArray(ast.NumberTerm("2"), ast.NumberTerm("2")),
+		ast.NewArray(ast.NumberTerm("2"), ast.NumberTerm("1")),
+	}
+
+	if len(tracer) != len(expected) {
+		t.Fatalf("unexpected result length %d", len(tracer))
+	}
+
+	for i, value := range tracer {
+		if value.Compare(expected[i]) != 0 {
+			t.Errorf("%v != %v at index %d", value, expected[i], i)
+		}
+	}
+}
+
+type plugRuleHeadKeyRecorder []ast.Value
+
+func (pr plugRuleHeadKeyRecorder) Enabled() bool {
```

Comment:
[nit]
```suggestion
func (plugRuleHeadKeyRecorder) Enabled() bool {
```
(also line 1030, but it really doesn't matter much)